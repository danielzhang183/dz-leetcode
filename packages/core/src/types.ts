export interface CodeSnippet {
  lang: string
  code: string
}

export interface TestCase {
  toBe: string
  expect: string
}

export interface TopicTag {
  slug: string
  translatedName: string
}

export type Category =
  | 'algorithm'
  | 'structure'
  | 'interview'
  | 'design'

export type AlgorithmTag =
  | 'back-tracking' | 'bt'
  | 'binary-search' | 'bs'
  | 'dynamic-programming' | 'dp'
  | 'greedy'
  | 'math'
  | 'sort'

export type StructureTag =
  | 'hash-table' | 'ht'
  | 'linked-list' | 'll'
  | 'array'
  | 'heap'
  | 'queue'
  | 'stack'
  | 'tree'

export type Tag = AlgorithmTag | StructureTag

export type Difficulty = 'Easy' | 'Medium' | 'Hard'

export interface Question {
  name: string
  title: string
  id: number | string
  difficulty: Difficulty
  category: Category | string
  tag: Tag
  link: string
  origin: string
}

export interface RawQuestion {
  questionId: string
  categoryTitle: string
  title: string
  titleSlug: string
  content: string
  difficulty: Difficulty
  exampleTestcases: string
  codeSnippets: Array<CodeSnippet>
  topicTags: Array<TopicTag>
}

export interface ResolvedQuestion {
  questionId: string
  category: string
  title: string
  titleSlug: string
  content: string
  difficulty: Difficulty
  testcases: Array<TestCase>
  code: string
  functionName: string
  tag: Tag | string
  path: string
  origin: string
}

export interface WritableQuestions {
  questions: Array<Question>
}

export interface GenerateOptions {
  category?: string
  tag?: string
  identifier: string | number | undefined
}

export interface GenerateReturn {
  type: 'markdown' | 'category' | 'code' | 'testcase'
  outFile: string
  content: string
}

export interface GenerateError {
  category?: string
  tag?: string
  error: unknown
  timestamp: number
}

export interface ErrorLogBase {
  timestamp: number
  error: unknown
}

export interface BatchErrorLog extends ErrorLogBase {
  type: 'batch-error'
  file: string
}

export interface SingleErrorLog extends ErrorLogBase {
  type: 'single-error'
  question?: string | number
}

export type RuntimeErrorLog = BatchErrorLog | SingleErrorLog
