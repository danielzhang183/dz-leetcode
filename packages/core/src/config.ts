import { resolve } from 'path'
import { createConfigLoader } from 'unconfig'
import createDebug from 'debug'
import deepmerge from 'deepmerge'
import { isString, toArray } from './utils'
import type { CommonOptions } from './types'

const debug = createDebug('dz-leetcode:config')

export function normalizeConfig<T extends CommonOptions>(options: T): T {
  const {
    root = process.cwd(),
    paths = {},
  } = options

  const { code, doc } = paths
  paths.code = resolve(root, code || './packages/code')
  paths.doc = resolve(root, doc || './packages/docs')

  options.root = root
  options.paths = paths
  options.lang = toArray(options.lang).flatMap(i => i.split(','))
  options.identifier = toArray(options.identifier).flatMap(i => (isString(i) && i.split(',')) || i)
  options.isResolved = true

  return options
}

export async function resolveConfig<T extends CommonOptions>(options: T): Promise<T> {
  if (options.isResolved)
    return options

  options = normalizeConfig(options)

  const loader = createConfigLoader<CommonOptions>({
    sources: [
      {
        files: [
          'leetcode.config',
        ],
      },
      {
        files: [
          '.leetcoderc',
        ],
        extensions: ['json', ''],
      },
    ],
    cwd: options.cwd || process.cwd(),
    merge: false,
  })

  const config = await loader.load()

  if (!config.sources.length)
    return options

  debug(`config file found ${config.sources[0]}`)
  const configOptions = normalizeConfig(config.config)

  return deepmerge(options, configOptions) as T
}
