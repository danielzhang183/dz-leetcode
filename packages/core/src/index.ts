export * from './generate'
export * from './question'
export * from './types'
