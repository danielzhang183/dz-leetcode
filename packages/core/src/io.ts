import { existsSync, promises as fs } from 'fs'
import { dirname } from 'path'
import YAML from 'js-yaml'
import type { WritableQuestions } from './types'

export function parse<T>(raw: string): T | undefined {
  raw = raw.trim()
  if (!raw)
    return

  return (raw.startsWith('['))
    ? JSON.parse(raw)
    : YAML.load(raw)
}

export function stringify(obj: WritableQuestions, useYaml = true) {
  if (!Object.keys(obj).length)
    return

  return useYaml
    ? YAML.dump(obj, { indent: 2 }).trimEnd()
    : Object.keys(obj).length > 1
      ? JSON.stringify(obj, null, 2)
      : JSON.stringify(obj)
}

export async function readFile(file: string) {
  if (!existsSync(file))
    return

  return await fs.readFile(file, 'utf-8')
}

export async function writeFile(outFile: string, content: string) {
  const dir = dirname(outFile)
  if (!existsSync(dir))
    await fs.mkdir(dir, { recursive: true })
  await fs.writeFile(outFile, content, 'utf-8')
}
