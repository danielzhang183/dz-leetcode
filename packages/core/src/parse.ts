import YAML from 'js-yaml'
import type { WritableQuestions } from './types'

export function parse<T>(raw: string): T | undefined {
  raw = raw.trim()
  if (!raw)
    return undefined
  return (raw.startsWith('['))
    ? JSON.parse(raw)
    : YAML.load(raw)
}

export function stringify(obj: WritableQuestions, useYaml = true) {
  if (!Object.keys(obj).length)
    return undefined
  return useYaml
    ? YAML.dump(obj, { indent: 2 }).trimEnd()
    : Object.keys(obj).length > 1
      ? JSON.stringify(obj, null, 2)
      : JSON.stringify(obj)
}
