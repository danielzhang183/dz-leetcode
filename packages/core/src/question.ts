import { join } from 'path'
import { $fetch } from 'ohmyfetch'
import { hyphenate, pad } from './utils'
import type { Category, CodeSnippet, Question, RawQuestion, ResolvedQuestion, Tag, TestCase, TopicTag } from './types'

const LEETCODE_FETCH_URL = 'https://leetcode.cn/graphql/'
const LEETCODE_QUESTION_URL = 'https://leetcode.cn/problems'
const LEETCODE_LIST_QUERY = [
  'query problemsetQuestionList($categorySlug: String, $limit: Int, $skip: Int, $filters: QuestionListFilterInput) {',
  '  problemsetQuestionList(',
  '    categorySlug: $categorySlug',
  '    limit: $limit',
  '    skip: $skip',
  '    filters: $filters',
  '  ) {',
  '    hasMore',
  '    total',
  '    questions {',
  '      difficulty',
  '      frontendQuestionId',
  '      titleSlug',
  '    }',
  '  }',
  '}',
].join('\n')
const LEETCODE_QUESTION_QUERY = [
  'query questionData($titleSlug: String!) {',
  '  question(titleSlug: $titleSlug) {',
  '    questionId',
  '    categoryTitle',
  '    title',
  '    titleSlug',
  '    content',
  '    difficulty',
  '    exampleTestcases',
  '    codeSnippets {',
  '      lang',
  '      code',
  '    }',
  '    topicTags {',
  '      slug',
  '       translatedName',
  '    }',
  '  }',
  '}',
].join('\n')
const LEETCODE_TAGS_QUERY = [
  'query questionTagTypeWithTags {',
  '  questionTagTypeWithTags {',
  '    name',
  '    transName',
  '    tagRelation {',
  '      questionNum',
  '      tag {',
  '        name',
  '        id',
  '        nameTranslated',
  '        slug',
  '      }',
  '    }',
  '  }',
  '}',
].join('\n')

export interface FilterOptions {
  category?: string
  tag?: string
  difficulty?: string
  skip?: number
}

export interface ResolvedFilterOptions {
  listId?: string
  tags?: string[]
  difficulty?: string
}

export interface RandomQuestion {
  difficulty: string
  questionId: string
  titleSlug: string
}

export interface BaseMeta {
  name: string
  nameTranslated: string
}

export interface TagMeta extends BaseMeta {
  id: string
  slug: string
}

export interface CategoryMap extends BaseMeta {
  tagMap: Array<{
    questionNum: number
    tag: TagMeta
  }>
}

export function getCategoryMaps(): Promise<CategoryMap[]> {
  return $fetch(LEETCODE_FETCH_URL, {
    method: 'post',
    body: {
      query: LEETCODE_TAGS_QUERY,
      variables: {},
    },
  }).then(
    r => r.data.questionTagTypeWithTags.map((i: any): CategoryMap => ({
      name: i.name,
      nameTranslated: i.transName,
      tagMap: i.tagRelation,
    })) || [],
  )
}

export function getQuestionsByFilter(options: FilterOptions): Promise<RandomQuestion[]> {
  const {
    category,
    tag,
    difficulty,
    skip = 0,
  } = options

  const filters: ResolvedFilterOptions = {}
  if (category)
    filters.listId = category
  if (tag)
    filters.tags = tag.split(',')
  if (difficulty)
    filters.difficulty = difficulty.toUpperCase()

  return $fetch(LEETCODE_FETCH_URL, {
    method: 'post',
    body: {
      query: LEETCODE_LIST_QUERY,
      variables: {
        categorySlug: '',
        skip,
        limit: 100,
        filters: {
          orderBy: 'FRONTEND_ID',
          sortOrder: 'DESCENDING',
          ...filters,
        },
      },
    },
  }).then(
    r => r.data?.problemsetQuestionList?.questions.map((i: any): RandomQuestion => ({
      difficulty: i.difficulty,
      titleSlug: i.titleSlug,
      questionId: i.frontendQuestionId,
    })) || [],
  )
}

export async function getQuestionById(id: number): Promise<RawQuestion | undefined> {
  const titleSlug = await $fetch(LEETCODE_FETCH_URL, {
    method: 'post',
    body: {
      query: LEETCODE_LIST_QUERY,
      variables: {
        categorySlug: '',
        skip: id - 1,
        limit: 1,
        filters: {},
      },
    },
  }).then(r => r.data.problemsetQuestionList.questions[0]?.titleSlug)

  return titleSlug ? await getQuestionByTitle(titleSlug) : undefined
}

export function getQuestionByTitle(titleSlug: string): Promise<RawQuestion | undefined> {
  return $fetch(LEETCODE_FETCH_URL, {
    method: 'post',
    body: {
      operationName: 'questionData',
      query: LEETCODE_QUESTION_QUERY,
      variables: { titleSlug },
    },
  }).then(r => r.data?.question || undefined)
}

export const getQuestionPath = (
  category: string,
  tag: Tag | string,
  questionId: string | number,
  ext?: '.md' | '.ts' | '.test.ts' | '.yml' | '.yaml' | '.json',
) => join(
  hyphenate(category),
  hyphenate(tag),
  `${pad(questionId)}${ext ?? ''}`,
)

export const getQuestionOrigin = (titleSlug: string) => `${LEETCODE_QUESTION_URL}/${titleSlug}`

export function normalizeRawQuestion(
  question: RawQuestion,
  category?: Category | string | undefined,
  tag?: Tag | string | undefined,
): ResolvedQuestion {
  function normalizeTestCases(exampleTestcases: string): TestCase[] {
    const res: TestCase[] = []
    const testcases = exampleTestcases.split('\n')
    for (let i = 0; i < testcases.length - 1; i += 2)
      res.push({ expect: testcases[i], toBe: testcases[i + 1] })

    return res
  }

  function normalizeCode(codeSnippets: CodeSnippet[]): string {
    return codeSnippets?.find(i => i.lang === 'TypeScript')?.code || ''
  }

  function normalizeFunctionName(code: string): string {
    const re = /function \*?(\w+)\(/
    return code?.match(re)![1] || 'fn'
  }

  function normalizeTag(tag: string | TopicTag[]): string {
    return hyphenate(Array.isArray(tag) ? tag[0].slug : tag)
  }

  const code = normalizeCode(question.codeSnippets)
  const normalizedCategory = hyphenate(category || question.categoryTitle)
  const normalizedTag = normalizeTag(tag || question.topicTags)

  return {
    questionId: question.questionId,
    category: normalizedCategory,
    tag: normalizedTag,
    title: question.title,
    titleSlug: question.titleSlug,
    content: question.content,
    difficulty: question.difficulty,
    testcases: normalizeTestCases(question.exampleTestcases),
    code,
    functionName: normalizeFunctionName(code),
    path: getQuestionPath(normalizedCategory, normalizedTag, question.questionId),
    origin: getQuestionOrigin(question.titleSlug),
    outFiles: [],
  }
}

export function normalizeResolvedQuestion(question: ResolvedQuestion): Question {
  return {
    name: question.title,
    title: question.titleSlug,
    difficulty: question.difficulty,
    id: Number(question.questionId),
    link: `/${question.path}`,
    origin: question.origin,
    tag: question.tag as Tag,
    category: question.category,
  }
}
