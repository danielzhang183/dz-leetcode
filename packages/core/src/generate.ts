import { join, resolve } from 'path'
import MagicString from 'magic-string'
import { getQuestionById, getQuestionByTitle, normalizeRawQuestion, normalizeResolvedQuestion } from './question'
import type { GenerateReturn, ImportableQuestionOptions, ResolvedQuestion, WritableQuestions } from './types'
import { readFile, writeFile } from './utils'
import { parse, stringify } from './parse'

export const root = resolve(process.cwd(), './packages')
export const pathCode = join(root, './code')
export const pathDocs = join(root, './docs')

export function genMarkdown(question: ResolvedQuestion): GenerateReturn {
  const {
    code,
    content,
    questionId,
    title,
    difficulty,
    origin,
    path,
  } = question

  function genMarkdownPre(s: MagicString) {
    [
      '---',
      `title: ${title}`,
      `display: ${questionId}. ${title}`,
      `difficulty: ${difficulty}`,
      `origin: ${origin}`,
      '---',
      '\n[[toc]]\n',
      '## Problem\n',
    ]
      .reverse()
      .forEach(str => s.prepend(`${str}\n`))
  }

  function genMarkdownPost(s: MagicString) {
    [
      '\n\n## Solution\n',
      '```ts',
      code,
      '```\n',
      `[view source](${origin})`,
    ]
      .forEach(str => s.append(`${str}\n`))
  }

  function transformMarkdown(s: MagicString) {
    s
      .replaceAll('&nbsp;', ' ')
      .replaceAll(/\t/g, '  ')
      .replaceAll(/<\/?p>/g, '')
      .replaceAll(/<\/?code>/g, '')
      .replaceAll(/<strong\sclass=.*?>/g, '<strong>')
  }

  const s = new MagicString(content)
  transformMarkdown(s)
  genMarkdownPre(s)
  genMarkdownPost(s)

  return {
    type: 'markdown',
    outFile: join(pathDocs, 'src/pages', `${path}.md`),
    content: s.toString(),
  }
}

export async function genCatelog(question: ResolvedQuestion): Promise<GenerateReturn> {
  const path = join(pathDocs, 'data', `${question.path.replace(/(\/\d+$)/g, '')}.yml`)
  const content = parse<WritableQuestions>(await readFile(path) || '') || { questions: [] }
  if (!content.questions.find(i => i.title === question.titleSlug))
    content.questions.push(normalizeResolvedQuestion(question))

  return {
    type: 'category',
    outFile: path,
    content: stringify(content)!,
  }
}

export function genCode(question: ResolvedQuestion): GenerateReturn {
  const { code, path } = question

  return {
    type: 'code',
    outFile: join(pathCode, 'src', `${path}.ts`),
    content: `export ${code}`,
  }
}

export function genTestCase(question: ResolvedQuestion): GenerateReturn {
  const { testcases, functionName, path } = question

  const s = new MagicString('')
  ;[
    'import { describe, expect, it } from \'vitest\'',
    `import { ${functionName} } from '../../../src/${path}'\n`,
    `describe('${functionName}', () => {`,
    '  it(\'exported\', () => {',
    ...testcases.map(
      ({ expect, toBe }) => `    expect(${functionName}(${expect})).toBe(${toBe})`,
    ),
    '  })',
    '})',
  ]
    .forEach(str => s.append(`${str}\n`))

  return {
    type: 'testcase',
    outFile: join(pathCode, 'test', `${path}.test.ts`),
    content: s.toString(),
  }
}

// export async function generate(file: string): Promise<RuntimeErrorLog[]> {
//   const questions = parse<ImportableQuestions>(await readFile(file) || '')?.questions
//   const errorLogs: RuntimeErrorLog[] = []
//   if (!questions || !questions.length) {
//     errorLogs.push({
//       type: 'batch-error',
//       file,
//       timestamp: Date.now(),
//       error: `${file} has no import data!`,
//     })

//     return errorLogs
//   }

//   async function batchGenerate(questions: ImportableQuestionOptions[]) {
//     const logs = await Promise.all(
//       questions.map(async (question: ImportableQuestionOptions): Promise<SingleErrorLog | ResolvedQuestion> => {
//         return await singleGenerate(question)
//       }))
//     return logs.filter(i => i) as SingleErrorLog[]
//   }

//   return batchGenerate(questions)
// }

export async function singleGenerate(options: ImportableQuestionOptions): Promise<{
  error: any
  question: ResolvedQuestion | null
}> {
  const { category, tag, id, name } = options
  if (!id && !name) {
    return {
      question: null,
      error: 'Give question name or id at least',
    }
  }

  const rawQuestion = id
    ? await getQuestionById(id)
    : await getQuestionByTitle(name!)
  if (!rawQuestion) {
    return {
      question: null,
      error: `Question ${id ? `No.${id}` : name} Not Found!`,
    }
  }

  const question = normalizeRawQuestion(rawQuestion, category, tag)
  const gens = await Promise.all([
    genMarkdown(question),
    genCatelog(question),
    genCode(question),
    genTestCase(question),
  ])
  await Promise.all(gens.map(async ({ outFile, content }) => {
    await writeFile(outFile, content)
  }))

  return {
    question,
    error: null,
  }
}
