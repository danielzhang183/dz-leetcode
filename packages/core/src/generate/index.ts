import { resolveConfig } from '../config'
import { writeFile } from '../io'
import { getQuestionById, getQuestionByTitle, normalizeRawQuestion } from '../question'
import type { CommonOptions, GenerateError, GenerateOptions, GenerateReturn, ResolvedQuestion } from '../types'
import { isNumber } from '../utils'
import { genCatalog, genCode, genMarkdown, genTestCase } from './gen'

export async function generate(options: GenerateOptions): Promise<GenerateReturn> {
  options = await resolveConfig(options)

  const {
    category,
    tag,
    identifier,
    lang,
    write = true, // write to file or not
    isTranslated = false, // translate to zh_CN or not
  } = options
  if (!identifier) {
    return {
      error: generateError('Give question name or id at least', category, tag),
    }
  }

  const rawQuestion = isNumber(identifier)
    ? await getQuestionById(identifier)
    : await getQuestionByTitle(identifier)
  if (!rawQuestion) {
    return {
      error: generateError(`Question ${identifier} Not Found!`, category, tag),
    }
  }

  const question = normalizeRawQuestion(
    rawQuestion,
    { category, tag, lang, isTranslated },
  )
  if (write)
    return await writeQuestion(question, options)

  return { question }
}

export function generateError(error: unknown, category = 'unknown-category', tag = 'unknown-tag'): GenerateError {
  return {
    category,
    tag,
    error,
    timestamp: Date.now(),
  }
}

export async function writeQuestion<T extends CommonOptions>(question: ResolvedQuestion, options: T): Promise<GenerateReturn> {
  const {
    onlyDoc = false,
    paths = {},
  } = options

  const gens = [
    genMarkdown(question, paths.doc),
    genCatalog(question, paths.doc),
  ]
  if (!onlyDoc) {
    gens.push(...[
      genCode(question, paths.code),
      genTestCase(question, paths.code),
    ])
  }

  try {
    // wait generate content for ready
    const resolvedGens = await Promise.all(gens)
    // write generate content to its' relative file
    const outFiles = await Promise.all(
      resolvedGens.map(async ({ outFile, content }) => {
        await writeFile(outFile, content)
        return outFile
      }),
    )
    question.outFiles = outFiles.sort()
  }
  catch (error) { // TODO: not handle carefully
    return {
      error: generateError(error, question.category, question.tag),
    }
  }

  return { question }
}
