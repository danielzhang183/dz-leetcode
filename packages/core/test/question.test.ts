import { describe, expect, it } from 'vitest'
import type { FilterOptions } from '../src'
import { getQuestionById, getQuestionsByFilter } from '../src'
import { question, rawQuestion, resolvedQuestion } from './fixture'

describe('question', () => {
  it('rawQuestion', async () => {
    expect(rawQuestion).toMatchSnapshot('rawQuestion')
  })

  it('resolvedQuestion', () => {
    expect(resolvedQuestion).toMatchSnapshot('resolvedQuestion')
  })

  it('question', () => {
    expect(question).toMatchSnapshot('question')
  })

  it('get question by id', async () => {
    expect(await getQuestionById(1)).toStrictEqual(rawQuestion)
  })
})

describe('get questions by filter', () => {
  const makeOptions = (category?: string, tag?: string, difficulty?: string): FilterOptions => {
    const options: FilterOptions = {}
    if (category)
      options.category = category
    if (tag)
      options.tag = tag
    if (difficulty)
      options.difficulty = difficulty.toUpperCase()

    return options
  }

  it('with category', async () => {
    expect(await (await getQuestionsByFilter(makeOptions('binary-search'))).slice(0, 4)).toMatchInlineSnapshot(`
      [
        {
          "difficulty": "MEDIUM",
          "questionId": "1901",
          "titleSlug": "find-a-peak-element-ii",
        },
        {
          "difficulty": "MEDIUM",
          "questionId": "1898",
          "titleSlug": "maximum-number-of-removable-characters",
        },
        {
          "difficulty": "MEDIUM",
          "questionId": "1894",
          "titleSlug": "find-the-student-that-will-replace-the-chalk",
        },
        {
          "difficulty": "MEDIUM",
          "questionId": "1891",
          "titleSlug": "cutting-ribbons",
        },
      ]
    `)

    const result1 = await getQuestionsByFilter(makeOptions('', '', 'easy'))
    expect(result1.slice(0, 4).every(i => i.difficulty === 'EASY')).toBe(true)
  })
})
