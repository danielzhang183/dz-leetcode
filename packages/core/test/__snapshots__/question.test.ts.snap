// Vitest Snapshot v1

exports[`question > question > question 1`] = `
{
  "category": "algorithms",
  "difficulty": "Easy",
  "id": 1,
  "link": "/algorithms/array/001",
  "name": "Two Sum",
  "origin": "https://leetcode.cn/problems/two-sum",
  "tag": "array",
  "title": "two-sum",
}
`;

exports[`question > rawQuestion > rawQuestion 1`] = `
{
  "categoryTitle": "Algorithms",
  "codeSnippets": [
    {
      "code": "class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        
    }
};",
      "lang": "C++",
    },
    {
      "code": "class Solution {
    public int[] twoSum(int[] nums, int target) {

    }
}",
      "lang": "Java",
    },
    {
      "code": "class Solution(object):
    def twoSum(self, nums, target):
        \\"\\"\\"
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        \\"\\"\\"",
      "lang": "Python",
    },
    {
      "code": "class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:",
      "lang": "Python3",
    },
    {
      "code": "

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* twoSum(int* nums, int numsSize, int target, int* returnSize){

}",
      "lang": "C",
    },
    {
      "code": "public class Solution {
    public int[] TwoSum(int[] nums, int target) {

    }
}",
      "lang": "C#",
    },
    {
      "code": "/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {

};",
      "lang": "JavaScript",
    },
    {
      "code": "# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[]}
def two_sum(nums, target)

end",
      "lang": "Ruby",
    },
    {
      "code": "class Solution {
    func twoSum(_ nums: [Int], _ target: Int) -> [Int] {

    }
}",
      "lang": "Swift",
    },
    {
      "code": "func twoSum(nums []int, target int) []int {

}",
      "lang": "Go",
    },
    {
      "code": "object Solution {
    def twoSum(nums: Array[Int], target: Int): Array[Int] = {

    }
}",
      "lang": "Scala",
    },
    {
      "code": "class Solution {
    fun twoSum(nums: IntArray, target: Int): IntArray {

    }
}",
      "lang": "Kotlin",
    },
    {
      "code": "impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {

    }
}",
      "lang": "Rust",
    },
    {
      "code": "class Solution {

    /**
     * @param Integer[] $nums
     * @param Integer $target
     * @return Integer[]
     */
    function twoSum($nums, $target) {

    }
}",
      "lang": "PHP",
    },
    {
      "code": "function twoSum(nums: number[], target: number): number[] {

};",
      "lang": "TypeScript",
    },
    {
      "code": "(define/contract (two-sum nums target)
  (-> (listof exact-integer?) exact-integer? (listof exact-integer?))

  )",
      "lang": "Racket",
    },
    {
      "code": "-spec two_sum(Nums :: [integer()], Target :: integer()) -> [integer()].
two_sum(Nums, Target) ->
  .",
      "lang": "Erlang",
    },
    {
      "code": "defmodule Solution do
  @spec two_sum(nums :: [integer], target :: integer) :: [integer]
  def two_sum(nums, target) do

  end
end",
      "lang": "Elixir",
    },
    {
      "code": "class Solution {
  List<int> twoSum(List<int> nums, int target) {

  }
}",
      "lang": "Dart",
    },
  ],
  "content": "<p>Given an array of integers <code>nums</code>&nbsp;and an integer <code>target</code>, return <em>indices of the two numbers such that they add up to <code>target</code></em>.</p>

<p>You may assume that each input would have <strong><em>exactly</em> one solution</strong>, and you may not use the <em>same</em> element twice.</p>

<p>You can return the answer in any order.</p>

<p>&nbsp;</p>
<p><strong class=\\"example\\">Example 1:</strong></p>

<pre>
<strong>Input:</strong> nums = [2,7,11,15], target = 9
<strong>Output:</strong> [0,1]
<strong>Explanation:</strong> Because nums[0] + nums[1] == 9, we return [0, 1].
</pre>

<p><strong class=\\"example\\">Example 2:</strong></p>

<pre>
<strong>Input:</strong> nums = [3,2,4], target = 6
<strong>Output:</strong> [1,2]
</pre>

<p><strong class=\\"example\\">Example 3:</strong></p>

<pre>
<strong>Input:</strong> nums = [3,3], target = 6
<strong>Output:</strong> [0,1]
</pre>

<p>&nbsp;</p>
<p><strong>Constraints:</strong></p>

<ul>
	<li><code>2 &lt;= nums.length &lt;= 10<sup>4</sup></code></li>
	<li><code>-10<sup>9</sup> &lt;= nums[i] &lt;= 10<sup>9</sup></code></li>
	<li><code>-10<sup>9</sup> &lt;= target &lt;= 10<sup>9</sup></code></li>
	<li><strong>Only one valid answer exists.</strong></li>
</ul>

<p>&nbsp;</p>
<strong>Follow-up:&nbsp;</strong>Can you come up with an algorithm that is less than&nbsp;<code>O(n<sup>2</sup>)&nbsp;</code>time complexity?",
  "difficulty": "Easy",
  "exampleTestcases": "[2,7,11,15]
9
[3,2,4]
6
[3,3]
6",
  "questionId": "1",
  "title": "Two Sum",
  "titleSlug": "two-sum",
  "topicTags": [
    {
      "slug": "array",
      "translatedName": "数组",
    },
    {
      "slug": "hash-table",
      "translatedName": "哈希表",
    },
  ],
}
`;

exports[`question > resolvedQuestion > resolvedQuestion 1`] = `
{
  "category": "algorithms",
  "code": "function twoSum(nums: number[], target: number): number[] {

};",
  "content": "<p>Given an array of integers <code>nums</code>&nbsp;and an integer <code>target</code>, return <em>indices of the two numbers such that they add up to <code>target</code></em>.</p>

<p>You may assume that each input would have <strong><em>exactly</em> one solution</strong>, and you may not use the <em>same</em> element twice.</p>

<p>You can return the answer in any order.</p>

<p>&nbsp;</p>
<p><strong class=\\"example\\">Example 1:</strong></p>

<pre>
<strong>Input:</strong> nums = [2,7,11,15], target = 9
<strong>Output:</strong> [0,1]
<strong>Explanation:</strong> Because nums[0] + nums[1] == 9, we return [0, 1].
</pre>

<p><strong class=\\"example\\">Example 2:</strong></p>

<pre>
<strong>Input:</strong> nums = [3,2,4], target = 6
<strong>Output:</strong> [1,2]
</pre>

<p><strong class=\\"example\\">Example 3:</strong></p>

<pre>
<strong>Input:</strong> nums = [3,3], target = 6
<strong>Output:</strong> [0,1]
</pre>

<p>&nbsp;</p>
<p><strong>Constraints:</strong></p>

<ul>
	<li><code>2 &lt;= nums.length &lt;= 10<sup>4</sup></code></li>
	<li><code>-10<sup>9</sup> &lt;= nums[i] &lt;= 10<sup>9</sup></code></li>
	<li><code>-10<sup>9</sup> &lt;= target &lt;= 10<sup>9</sup></code></li>
	<li><strong>Only one valid answer exists.</strong></li>
</ul>

<p>&nbsp;</p>
<strong>Follow-up:&nbsp;</strong>Can you come up with an algorithm that is less than&nbsp;<code>O(n<sup>2</sup>)&nbsp;</code>time complexity?",
  "difficulty": "Easy",
  "functionName": "twoSum",
  "lang": "typescript",
  "origin": "https://leetcode.cn/problems/two-sum",
  "outFiles": [],
  "path": "algorithms/array/001",
  "questionId": "1",
  "tag": "array",
  "testcases": [
    {
      "expect": "[2,7,11,15]",
      "toBe": "9",
    },
    {
      "expect": "[3,2,4]",
      "toBe": "6",
    },
    {
      "expect": "[3,3]",
      "toBe": "6",
    },
  ],
  "title": "Two Sum",
  "titleSlug": "two-sum",
}
`;
